File: userservice/src/main/java/hu/bme/aut/classifiedadvertisementsite/userservice/security/UserDetailsImpl.java
package hu.bme.aut.classifiedadvertisementsite.userservice.security;

import com.fasterxml.jackson.annotation.JsonIgnore;
import hu.bme.aut.classifiedadvertisementsite.userservice.model.User;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

public class UserDetailsImpl implements UserDetails {

    private Integer id;

    private String username;

    private String email;

    @JsonIgnore
    private String password;

    private Boolean enabled;

    private Collection<? extends GrantedAuthority> authorities;

    public UserDetailsImpl(Integer id, String username, String email, String password, Boolean enabled, List<String> roles) {
        List<SimpleGrantedAuthority> authorities =
                roles.stream().map(role -> new SimpleGrantedAuthority(role)).toList();

        this.id = id;
        this.username = username;
        this.email = email;
        this.password = password;
        this.enabled = enabled;
        this.authorities = authorities;
    }

    public static UserDetailsImpl build(User user) {
        List<String> authorities = user.getRoles().stream()
                .map(role -> role.getName().name())
                .collect(Collectors.toList());

        return new UserDetailsImpl(
                user.getId(),
                user.getUsername(),
                user.getEmail(),
                user.getPassword(),
                user.getEnabled(),
                authorities);
    }

    public Integer getId() {
        return id;
    }

    public String getEmail() {
        return email;
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return authorities;
    }

    @Override
    public String getPassword() {
        return password;
    }

    @Override
    public String getUsername() {
        return username;
    }

    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return email != null;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return enabled;
    }
}

File: advertisementservice/src/main/kotlin/hu/bme/aut/classifiedadvertisementsite/advertisementservice/security/UserDetailsImpl.kt
package hu.bme.aut.classifiedadvertisementsite.advertisementservice.security

import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails


class UserDetailsImpl(
    private var id: Int,
    private var username: String,
    private var email: String,
    private var enabled: Boolean,
    roles: List<String?>
) : UserDetails {
    private var authorities: Collection<GrantedAuthority>? = null

    init {
        val authorities = roles.map { role: String? ->
            SimpleGrantedAuthority(role)
        }.toList()
        this.authorities = authorities
    }

    fun getId(): Int {
        return id
    }

    fun getEmail(): String {
        return email
    }

    override fun getAuthorities(): Collection<GrantedAuthority?>? {
        return authorities
    }

    override fun getPassword(): String? {
        return null
    }

    override fun getUsername(): String {
        return username
    }

    override fun isAccountNonExpired(): Boolean {
        return true
    }

    override fun isAccountNonLocked(): Boolean {
        return true
    }

    override fun isCredentialsNonExpired(): Boolean {
        return true
    }

    override fun isEnabled(): Boolean {
        return enabled
    }
}